farming/logics/traits/farming.rs

```diff
@@ -25,17 +30,112 @@ pub const MAX_PERIOD: u8 = 23u8;
 pub const FIRST_PERIOD_REWERD_SUPPLY: Balance = 151629858171523000000u128;
 
 #[openbrush::trait_definition]
-pub trait Farming: Storage<Data> + Storage<ownable::Data> + FarmingGetters {
+pub trait Farming: Storage<Data> + Storage<ownable::Data> + FarmingGetters + FarmingEvents {
     #[ink(message)]
     #[modifiers(only_owner)]
     fn add(
         &mut self,
-        alloc_point: u128,
+        alloc_point: u32,
         lp_token: AccountId,
-        rewarder: AccountId,
+        rewarder: Option<AccountId>,
     ) -> Result<(), FarmingError> {
         self._check_pool_duplicate(lp_token)?;
         self._update_all_pools()?;
+        self.data::<Data>().total_alloc_point = self
+            .data::<Data>()
+            .total_alloc_point
+            .checked_add(alloc_point)
+            .ok_or(FarmingError::AddOverflow2)?;
+        self.data::<Data>().lp_tokens.push(lp_token);
+        self.data::<Data>().rewarders.push(rewarder);
+        let pool_length = self.pool_length();
+
+        self.data::<Data>().pool_info.insert(
+            &pool_length,
+            &Pool {
+                acc_arsw_per_share: 0,
+                last_reward_block: Self::env().block_number(),
+                alloc_point,
+            },
+        );
+        self.data::<Data>().pool_info_length = pool_length
+            .checked_add(1)
+            .ok_or(FarmingError::AddOverflow2)?;
+        self._emit_log_pool_addition_event(pool_length, alloc_point, lp_token, rewarder);
+        Ok(())
+    }
+
+    #[ink(message)]
+    fn pending_arsw(&self, _pool_id: u32, _user: AccountId) -> Result<Balance, FarmingError> {
+        Ok(1_000_000_000_000_000_000)
+    }
+
+    #[ink(message)]
+    fn deposit(
+        &mut self,
+        pool_id: u32,
+        amount: Balance,
+        to: AccountId,
+    ) -> Result<(), FarmingError> {
+        let (prev_amount, prev_reward_debt) = self
+            .get_user_info(pool_id, to)
+            .ok_or(FarmingError::UserNotFound)?;
+        // TODO: Fix reward_debt
+        self.data::<Data>().user_info.insert(
+            &(pool_id, to),
+            &(
+                prev_amount
+                    .checked_add(amount)
+                    .ok_or(FarmingError::AddOverflow1)?,
+                prev_reward_debt,
+            ),
+        );
+        let lp_token = self
+            .get_lp_token(pool_id)
+            .ok_or(FarmingError::PoolNotFound2)?;
+        PSP22Ref::transfer_from(
```

Comment:
You should allow reentrancy for this or it will not work:
```
        PSP22Ref::transfer_from_builder(&lp_token, Self::env().caller(), Self::env().account_id(), value, Vec::<u8>::new())
            .call_flags(CallFlags::default().set_allow_reentry(true))
            .fire()
            .unwrap()?;
```

Reply:
Ah, I missed that! Thanks a lot. Will fix it

Reply:
Think we don't need to allow reentrancy. Pair contract only calls itself rather than transfer_from caller. Hence, we shouldn't allow reentrancy actually in this case.

Reply:
Yes I need to update in Pair contract. But here it is still relevant

Reply:
So, from the pair contract, you will callback the caller. But for what?

Reply:
In this case you want to transfer token from caller to this contract:
This contract calls PSP22 -> PSP22 will callback this contract in before_receive() -> reentrancy is happening.

In Router contract use transfer_builder but actually no reentrancy is possible so I'll fix that (and remove it)

Reply:
To be on the same page, from transfer action, internally `_do_safe_transfer_check` will be called, which in turn, will call back the caller. 
But for the farming contract, we haven't implemented `PSP22Receiver` trait deliberately as we don't need to safe checking feature.

So the conclusion I reached is we don't need to bother with reentrancy here. Am I wrong?

Reply:
`_do_safe_transfer_check` is part of `Internal` Trait of psp22 (not farming) and you should implement it in your psp22 to get the getters & events (and it will be implemented in standard psp22s as well)

Reply:
Ah, not the farming contract itself, the caller for `deposit` might call back the operator part of `before_received`: https://github.com/Supercolony-net/openbrush-contracts/blob/dc666f08c4e3bec2bf6057bb083cbf5223a28063/contracts/src/traits/psp22/psp22.rs#L160. So we need to allow reentrancy, is it correct?

Reply:
For the `_do_safe_transfer_check`, yes it is implemented by default. Thus usually, we don't need to implement it by ourselves.

Reply:
For the record, it works fine as it is.

Reply:
With which psp22 implementation?

Reply:
Since it is LP token, the pair contract is the implementation.

Reply:
It's working for Pair because transfer_from is overridden and there is no `_do_safe_transfer_check`
https://github.com/AstarNetwork/wasm-showcase-dapps/blob/9aa96d24a9bd338141aa97f7d60e96958c692cfd/uniswap-v2/contracts/pair/lib.rs#L147

PSP22 implementation will impl the default Internal Trait - in this one there is `_do_safe_transfer_check` that will do a callback and will panic if we didn't allow reentrancy with flags.

Now you can still keep it like this (as LP are Pair contract) but it will not be compatible with standard psp22

Reply:
A note for myself: